### ISSUE TITLE
Compilation is  S  L  O  W

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#1]
Today, a fresh "configure.py --mode=release; ninja" takes a whopping 9 minutes on my machine.
All to often, chaning just one (header) file results in a large part of this being redone.
Life is too short for this :-)

We could try to improve these figures by reducing the inclusion of header files, moving more things to .cc files, not building all the tests by default, and perhaps other tricks.

P.S. as a reference to just how slow 9 minutes is, consider that a fresh compile of OSv on my machine takes just 4 minutes. Moreover in OSv, most changes require recompiling only a few files, and take just a few seconds.

**Comment by jacobgorm** [comment]
From reading this thread I can conclude that the afternoon I just invested trying to get scylladb to build (plus the rest of the evening it seems I have to wait for the build to complete) has been wasted. Excuse me for saying this, but people who cannot make a C++ build go faster than this should really stick to using python. I have never in my life seen a slower build, and believe me I have seen some pretty crappy ones in my time.

**Comment by avikivity** [comment]
Actually, we have reached the same conclusion, and Scylla 1.5 (due in about a month) will by written in Python 3.  Stay tuned!

**Comment by penberg** [comment]
@jacobgorm Thanks for the advice, we truly appreciate it!

As for build times, they're slow, but not unbearably slow…

```
[penberg@nero scylla]$ ./configure.py --mode=release --with=scylla
[penberg@nero scylla]$ time ninja-build -j8
[271/271] LINK build/release/scylla

real	14m0.755s
user	102m8.942s
sys	4m7.185s
```

**Comment by penberg** [comment]
...that's not to say that our build doesn't suck -- it does. There's tons of over eager header dependency issues to tackle, for example.

**Comment by jacobgorm** [comment]
It looks like some autogenerated (antlr?) header files is getting sucked in almost everywhere. Never before have I seen gcc consume >2GiB RAM per file it compiles. I have another C++ build (closed source) that I am working on (a 3d engine that a colleague wrote) that has almost exactly the same # LOC and also builds with ninja. Yet it takes less than 10 seconds for a full build, and less than .1 second for an empty build, on the same machine where scylladb took more than an hour. The difference is probably that it uses no STL headers, and that recursive includes are banned, so I would say that in comparison the scylladb build is indeed unbearably slow. For the sake of your project (that I have now given up on using for a customer POC, I will revert to plain cassandra for the time being) I urge you to fix this (probably trivial) build bug, before it wears down morale and scares away more potential users or customers. Nothing kills productivity like a slow build does.

**Comment by avikivity** [comment]
The problem is not excessive headers or STL.  It is heavy use of templates and is directly derived from how the system works.  If you think it is trivial, please post a patch.

**Comment by nyh** [comment]
@avikivity is correct that templates are a big part of the problem, but it's not the entire problem: Another big problem for developers (rather than users who compile a single time), is header file spaghetti (most of the code is in headers, and the headers are very interdependent) so changing one source file causes almost the entire project to be recompiled.

**Comment by jacobgorm** [comment]
I normally use a script like this to find the worst headers:
```bash
for i in `git ls-files| egrep '\.hh$'`
do
    touch -r $i ref
    touch $i
    (
        dirty=`ninja  build/release/scylla  -n -d explain 2>&1 |grep dirty | wc -l`
        echo $dirty $i
    )

    touch -r ref $i
done
```

But in scylla's case they are almost all equally bad, top of the list:
221 idl/uuid.idl.hh
220 idl/keys.idl.hh
213 idl/query.idl.hh
207 hashing.hh
207 bytes.hh

...

and the bottom:

194 api/hinted_handoff.hh
194 api/gossiper.hh
194 api/failure_detector.hh
194 api/endpoint_snitch.hh
194 api/compaction_manager.hh
194 api/commitlog.hh
194 api/column_family.hh
194 api/collectd.hh
194 api/cache_service.hh
194 api/api_init.hh
194 api/api.hh

**Comment by avikivity** [comment]
`.idl.hh` are fake dependencies, they won't cause 221 files to be rebuilt (and aren't included by 221 files), just re-evaluated.  Try `touch idl/* && ninja build/release/scylla`; the compiler won't be invoked.

Note sure who the 194 includers of api/* are.

**Comment by jacobgorm** [comment]
I've deleted the comment, because the builid wasn't complete to begin with. Once I have a complete build again to start from I will repost with real numbers (but it might take a while to get there.)

**Comment by jacobgorm** [comment]
Updated comment:
Using this script:
```bash
for i in `git ls-files| egrep '\.hh$'`
do
    touch -r $i ref
    touch $i
    (
        dirty=`ninja  build/release/scylla  -n -d explain 2>&1 |grep dirty | wc -l`
        echo $dirty $i
    )

    touch -r ref $i
done
```
This is the list of top offenders (via sort -r -n):
206 hashing.hh
206 bytes.hh
198 log.hh
195 utils/allocation_strategy.hh
195 gc_clock.hh
194 utils/UUID.hh
194 utils/serialization.hh
194 utils/managed_bytes.hh
194 db_clock.hh
193 types.hh
193 to_string.hh
193 tombstone.hh
193 timestamp.hh
193 exceptions/exceptions.hh
193 db/write_type.hh
193 db/consistency_level_type.hh
193 cql_serialization_format.hh
193 atomic_cell.hh
183 unimplemented.hh
183 cql3/column_specification.hh
182 schema.hh
182 keys.hh
182 json.hh
182 database_fwd.hh
182 compress.hh
182 compound.hh
182 compound_compat.hh
182 compaction_strategy.hh
182 caching_options.hh
180 range.hh
...

In terms of of headers file triggering rebuilds.

**Comment by avikivity** [comment]
The list looks fairly sane.  Maybe compaction_strategy.hh can be replaced by forward declarations, but most of these headers are tiny, and are very stable.

**Comment by peldax** [comment]
I am sorry, but I completely agree with @jacobgorm . I have built bigger projects than this, but I cant built scylla from source. My computer has 6 Gb of RAM and it's not enough for compilation. Please dont tell me its sane or bearable. Its not.

**Comment by nyh** [comment]
I found an interesting project, https://github.com/aras-p/ClangBuildAnalyzer, which lets you analyze the output of clang's "-ftime-trace" to find the most expensive files, header files, templates, functions and a lot more.
I attach the full report for the "dev" build mode on my laptop, [report.txt](https://github.com/scylladb/scylla/files/7802138/report.txt)
and give some examples from it:

```
**** Time summary:
Compilation (848 times):
  Parsing (frontend):        10455.2 s
  Codegen & opts (backend):   4066.0 s
```

```
**** Files that took longest to parse (compiler frontend):
 89473 ms: build/dev/message/messaging_service.o
 85188 ms: build/dev/main.o
 71943 ms: build/dev/service/storage_service.o
 62647 ms: build/dev/service/storage_proxy.o
 62082 ms: build/dev/db/system_keyspace.o
 57676 ms: build/dev/api/storage_service.o
 57078 ms: build/dev/table.o
 56864 ms: build/dev/sstables/sstables.o
 55186 ms: build/dev/db/schema_tables.o

**** Files that took longest to codegen (compiler backend):
129226 ms: build/dev/message/messaging_service.o
110786 ms: build/dev/sstables/sstables.o
 98445 ms: build/dev/service/storage_proxy.o
 93259 ms: build/dev/service/storage_service.o
 92827 ms: build/dev/main.o
 71921 ms: build/dev/db/schema_tables.o
 71636 ms: build/dev/table.o
```
These list is not very helpful, because it doesn't help us figure out why these specific source files are so slow. However, it might (?) be worthwhile to split the longest-compiling ones to improve parallelism.

```
**** Templates that took longest to instantiate:
138239 ms: fmt::formatter<fmt::basic_string_view<char>>::format<fmt::basic_form... (351 times, avg 393 ms)
 85267 ms: utils::loading_cache<std::pair<seastar::basic_sstring<signed char, u... (118 times, avg 722 ms)
 84533 ms: seastar::throw_with_backtrace<marshal_exception, seastar::basic_sstr... (312 times, avg 270 ms)
 84404 ms: seastar::make_backtraced_exception_ptr<marshal_exception, seastar::b... (312 times, avg 270 ms)
```
I don't know what we can do about those. Maybe it makes sense to have external instantiation (only instantiating a specific version of a template) for some of these. Note that 138 seconds is almost 1% of the entire build time (around 15,000 seconds) - just for compiling this specific template 351 times :-(

```
**** Template sets that took longest to instantiate:
644482 ms: std::unique_ptr<$> (23125 times, avg 27 ms)
608400 ms: std::__and_<$> (370713 times, avg 1 ms)
522728 ms: std::__uniq_ptr_data<$> (23125 times, avg 22 ms)
516971 ms: std::__uniq_ptr_impl<$> (23125 times, avg 22 ms)
442828 ms: std::__or_<$> (352542 times, avg 1 ms)
361947 ms: fmt::format_to<$> (13824 times, avg 26 ms)
331887 ms: std::tuple<$> (33022 times, avg 10 ms)
318621 ms: seastar::format<$> (10721 times, avg 29 ms)
316114 ms: std::optional<$> (19862 times, avg 15 ms)
```
These are huge numbers (644 seconds for compiling std::unique_ptr 23,125 times), but other than crying about the design of C++, I'm not sure what we can do about it.

```
**** Functions that took longest to compile:
 10869 ms: cql3_parser::CqlParser::unaliasedSelector() (build/dev/gen/cql3/CqlParser.cpp)
  8101 ms: main::$_2::operator()() const::'lambda0'()::operator()() const (main.cc)
  6122 ms: database::setup_metrics() (database.cc)
  5854 ms: cql3::query_processor::query_processor(service::storage_proxy&, data... (cql3/query_processor.cc)
  5311 ms: sstables::init_metrics()::$_74::operator()() const (sstables/sstables.cc)
  5067 ms: cql3::functions::add_agg_functions(std::unordered_multimap<cql3::fun... (cql3/functions/aggregate_fcts.cc)
  4036 ms: alternator::stats::stats() (alternator/stats.cc)
```

Maybe we can take a look at some of these functions to figure out why they are so slow. For example, stats() is just 60 lines of code (!!!) but uses a lot of obviously super-unoptimized boost options tricks :-(

```
**** Function sets that took longest to compile / optimize:
 37473 ms: fmt::v7::detail::int_writer<$>::on_num() (1356 times, avg 27 ms)
 17437 ms: seastar::continuation<$>::run_and_dispose() (1978 times, avg 8 ms)
 13365 ms: seastar::future<$> seastar::future<$>::then_impl<$>(seastar::noncopy... (1471 times, avg 9 ms)
 12203 ms: seastar::noncopyable_function<$>::direct_vtable_for<$>::call(seastar... (964 times, avg 12 ms)
 10217 ms: seastar::internal::repeat_until_value_state<seastar::data_source_imp... (359 times, avg 28 ms)
 10209 ms: char const* fmt::v7::detail::parse_align<$>(char const*, char const*... (238 times, avg 42 ms)
```
I don't know what we can do about these.

```
** Expensive headers:
```
This is probably the most interesting section. For example
```
1232665 ms: seastar/include/seastar/core/smp.hh (included 329 times, avg 3746 ms), included via:
  storage_proxy.o storage_proxy.hh sharded.hh  (11982 ms)
  schema_registry.o sharded.hh  (9354 ms)
  connection_notifier.o connection_notifier.hh query_context.hh sharded.hh  (9206 ms)
  storage_service.o storage_service.hh sharded.hh  (9066 ms)
  data_listeners.o data_listeners.hh distributed.hh sharded.hh  (9061 ms)
  stream_manager.o distributed.hh sharded.hh  (8783 ms)
  query_processor.o query_processor.hh sharded.hh  (8636 ms)
  keyspace_utils.o keyspace_utils.hh sharded.hh  (8596 ms)
  event.o event.hh inet_address.hh ip.hh arp.hh net.hh  (7286 ms)
  untyped_result_set.o untyped_result_set.hh sharded.hh  (7096 ms)
  large_data_handler.o query_context.hh sharded.hh  (6743 ms)
  ...
```
According to this, almost 8% of the entire build time is spent on compiling seastar/core/smp.hh included 329 times. It's important to understand: 1. If we need to include it so much, 2. Why it is so slow (almost 4 seconds) for each include and if it can be made faster, 3. Maybe parts of it can be moved to a different header file to speed up the typical inclusions of smp.hh.

```
1086901 ms: types.hh (included 312 times, avg 3483 ms), included via:
  error_injection_fcts.o error_injection_fcts.hh native_scalar_function.hh native_function.hh abstract_function.hh function.hh  (13927 ms)
  type_json.o type_json.hh  (13269 ms)
  selector.o selector.hh assignment_testable.hh column_specification.hh  (13126 ms)
  user_function.o user_function.hh abstract_function.hh function.hh  (12823 ms)
  functions.o functions.hh function.hh  (12210 ms)
  mutation_utils.o mutation_utils.hh  (12179 ms)
  type_parser.o type_parser.hh  (12110 ms)
  values.o values.hh  (12067 ms)
  column_specification.o column_specification.hh  (11222 ms)
  castas_fcts.o castas_fcts.hh function.hh  (10970 ms)
  user_types.o user_types.hh column_specification.hh  (10927 ms)
  ...
```
Another 7% of the build time from including "types.hh" 312 times.

```
591047 ms: database.hh (included 74 times, avg 7987 ms), included via:
  view_update_generator.o view_update_generator.hh  (21102 ms)
  cql_type_parser.o  (20704 ms)
  table.o  (20328 ms)
  validation.o  (18638 ms)
  database.o  (17775 ms)
  view.o  (17062 ms)
  legacy_schema_migrator.o  (16223 ms)
  streams.o  (16097 ms)
  snapshot-ctl.o  (15231 ms)
  main.o  (14970 ms)
  sstables-format-selector.o  (14135 ms)
  ...
```
I think @avikivity has already been working on reducing the includes of database.hh, but it's still included 74 times - and each one is very slow (8 seconds) 
```
562433 ms: gms/inet_address.hh (included 291 times, avg 1932 ms), included via:
  event.o event.hh  (13498 ms)
  session_info.o session_info.hh  (12070 ms)
  gossiper.o  (12039 ms)
  storage_service.o storage_service.hh i_endpoint_state_change_subscriber.hh  (11846 ms)
  progress_info.o progress_info.hh  (11834 ms)
```
That is another header file - a rather silly one at that (how many source files need to handle IP addresses? Why does handling IP addresses take 2 seconds to compile?) included 291 times and single-handedly responsible for 3.7% of the build time. Heck, even seastar/core/future takes less of the build time (only slightly less, though).

**Comment by raphaelsc** [comment]
> I found an interesting project, https://github.com/aras-p/ClangBuildAnalyzer, which lets you analyze the output of clang's "-ftime-trace" to find the most expensive files, header files, templates, functions and a lot more. I attach the full report for the "dev" build mode on my laptop, [report.txt](https://github.com/scylladb/scylla/files/7802138/report.txt) and give some examples from it:
> 
> ```
> **** Time summary:
> Compilation (848 times):
>   Parsing (frontend):        10455.2 s
>   Codegen & opts (backend):   4066.0 s
> ```
> 
> ```
> **** Files that took longest to parse (compiler frontend):
>  89473 ms: build/dev/message/messaging_service.o
>  85188 ms: build/dev/main.o
>  71943 ms: build/dev/service/storage_service.o
>  62647 ms: build/dev/service/storage_proxy.o
>  62082 ms: build/dev/db/system_keyspace.o
>  57676 ms: build/dev/api/storage_service.o
>  57078 ms: build/dev/table.o
>  56864 ms: build/dev/sstables/sstables.o
>  55186 ms: build/dev/db/schema_tables.o
> 
> **** Files that took longest to codegen (compiler backend):
> 129226 ms: build/dev/message/messaging_service.o
> 110786 ms: build/dev/sstables/sstables.o
>  98445 ms: build/dev/service/storage_proxy.o
>  93259 ms: build/dev/service/storage_service.o
>  92827 ms: build/dev/main.o
>  71921 ms: build/dev/db/schema_tables.o
>  71636 ms: build/dev/table.o
> ```
> 
> These list is not very helpful, because it doesn't help us figure out why these specific source files are so slow. However, it might (?) be worthwhile to split the longest-compiling ones to improve parallelism.
> 
> ```
> **** Templates that took longest to instantiate:
> 138239 ms: fmt::formatter<fmt::basic_string_view<char>>::format<fmt::basic_form... (351 times, avg 393 ms)
>  85267 ms: utils::loading_cache<std::pair<seastar::basic_sstring<signed char, u... (118 times, avg 722 ms)
>  84533 ms: seastar::throw_with_backtrace<marshal_exception, seastar::basic_sstr... (312 times, avg 270 ms)
>  84404 ms: seastar::make_backtraced_exception_ptr<marshal_exception, seastar::b... (312 times, avg 270 ms)
> ```
> 
> I don't know what we can do about those. Maybe it makes sense to have external instantiation (only instantiating a specific version of a template) for some of these. Note that 138 seconds is almost 1% of the entire build time (around 15,000 seconds) - just for compiling this specific template 351 times :-(
> 
> ```
> **** Template sets that took longest to instantiate:
> 644482 ms: std::unique_ptr<$> (23125 times, avg 27 ms)
> 608400 ms: std::__and_<$> (370713 times, avg 1 ms)
> 522728 ms: std::__uniq_ptr_data<$> (23125 times, avg 22 ms)
> 516971 ms: std::__uniq_ptr_impl<$> (23125 times, avg 22 ms)
> 442828 ms: std::__or_<$> (352542 times, avg 1 ms)
> 361947 ms: fmt::format_to<$> (13824 times, avg 26 ms)
> 331887 ms: std::tuple<$> (33022 times, avg 10 ms)
> 318621 ms: seastar::format<$> (10721 times, avg 29 ms)
> 316114 ms: std::optional<$> (19862 times, avg 15 ms)
> ```
> 
> These are huge numbers (644 seconds for compiling std::unique_ptr 23,125 times), but other than crying about the design of C++, I'm not sure what we can do about it.
> 
> ```
> **** Functions that took longest to compile:
>  10869 ms: cql3_parser::CqlParser::unaliasedSelector() (build/dev/gen/cql3/CqlParser.cpp)
>   8101 ms: main::$_2::operator()() const::'lambda0'()::operator()() const (main.cc)
>   6122 ms: database::setup_metrics() (database.cc)
>   5854 ms: cql3::query_processor::query_processor(service::storage_proxy&, data... (cql3/query_processor.cc)
>   5311 ms: sstables::init_metrics()::$_74::operator()() const (sstables/sstables.cc)
>   5067 ms: cql3::functions::add_agg_functions(std::unordered_multimap<cql3::fun... (cql3/functions/aggregate_fcts.cc)
>   4036 ms: alternator::stats::stats() (alternator/stats.cc)
> ```
> 
> Maybe we can take a look at some of these functions to figure out why they are so slow. For example, stats() is just 60 lines of code (!!!) but uses a lot of obviously super-unoptimized boost options tricks :-(
> 
> ```
> **** Function sets that took longest to compile / optimize:
>  37473 ms: fmt::v7::detail::int_writer<$>::on_num() (1356 times, avg 27 ms)
>  17437 ms: seastar::continuation<$>::run_and_dispose() (1978 times, avg 8 ms)
>  13365 ms: seastar::future<$> seastar::future<$>::then_impl<$>(seastar::noncopy... (1471 times, avg 9 ms)
>  12203 ms: seastar::noncopyable_function<$>::direct_vtable_for<$>::call(seastar... (964 times, avg 12 ms)
>  10217 ms: seastar::internal::repeat_until_value_state<seastar::data_source_imp... (359 times, avg 28 ms)
>  10209 ms: char const* fmt::v7::detail::parse_align<$>(char const*, char const*... (238 times, avg 42 ms)
> ```
> 
> I don't know what we can do about these.
> 
> ```
> ** Expensive headers:
> ```
> 
> This is probably the most interesting section. For example
> 
> ```
> 1232665 ms: seastar/include/seastar/core/smp.hh (included 329 times, avg 3746 ms), included via:
>   storage_proxy.o storage_proxy.hh sharded.hh  (11982 ms)
>   schema_registry.o sharded.hh  (9354 ms)
>   connection_notifier.o connection_notifier.hh query_context.hh sharded.hh  (9206 ms)
>   storage_service.o storage_service.hh sharded.hh  (9066 ms)
>   data_listeners.o data_listeners.hh distributed.hh sharded.hh  (9061 ms)
>   stream_manager.o distributed.hh sharded.hh  (8783 ms)
>   query_processor.o query_processor.hh sharded.hh  (8636 ms)
>   keyspace_utils.o keyspace_utils.hh sharded.hh  (8596 ms)
>   event.o event.hh inet_address.hh ip.hh arp.hh net.hh  (7286 ms)
>   untyped_result_set.o untyped_result_set.hh sharded.hh  (7096 ms)
>   large_data_handler.o query_context.hh sharded.hh  (6743 ms)
>   ...
> ```
> 
> According to this, almost 8% of the entire build time is spent on compiling seastar/core/smp.hh included 329 times. It's important to understand: 1. If we need to include it so much, 2. Why it is so slow (almost 4 seconds) for each include and if it can be made faster, 3. Maybe parts of it can be moved to a different header file to speed up the typical inclusions of smp.hh.
> 
> ```
> 1086901 ms: types.hh (included 312 times, avg 3483 ms), included via:
>   error_injection_fcts.o error_injection_fcts.hh native_scalar_function.hh native_function.hh abstract_function.hh function.hh  (13927 ms)
>   type_json.o type_json.hh  (13269 ms)
>   selector.o selector.hh assignment_testable.hh column_specification.hh  (13126 ms)
>   user_function.o user_function.hh abstract_function.hh function.hh  (12823 ms)
>   functions.o functions.hh function.hh  (12210 ms)
>   mutation_utils.o mutation_utils.hh  (12179 ms)
>   type_parser.o type_parser.hh  (12110 ms)
>   values.o values.hh  (12067 ms)
>   column_specification.o column_specification.hh  (11222 ms)
>   castas_fcts.o castas_fcts.hh function.hh  (10970 ms)
>   user_types.o user_types.hh column_specification.hh  (10927 ms)
>   ...
> ```
> 
> Another 7% of the build time from including "types.hh" 312 times.
> 
> ```
> 591047 ms: database.hh (included 74 times, avg 7987 ms), included via:
>   view_update_generator.o view_update_generator.hh  (21102 ms)
>   cql_type_parser.o  (20704 ms)
>   table.o  (20328 ms)
>   validation.o  (18638 ms)
>   database.o  (17775 ms)
>   view.o  (17062 ms)
>   legacy_schema_migrator.o  (16223 ms)
>   streams.o  (16097 ms)
>   snapshot-ctl.o  (15231 ms)
>   main.o  (14970 ms)
>   sstables-format-selector.o  (14135 ms)
>   ...
> ```
> 
> I think @avikivity has already been working on reducing the includes of database.hh, but it's still included 74 times - and each one is very slow (8 seconds)
> 
> ```
> 562433 ms: gms/inet_address.hh (included 291 times, avg 1932 ms), included via:
>   event.o event.hh  (13498 ms)
>   session_info.o session_info.hh  (12070 ms)
>   gossiper.o  (12039 ms)
>   storage_service.o storage_service.hh i_endpoint_state_change_subscriber.hh  (11846 ms)
>   progress_info.o progress_info.hh  (11834 ms)
> ```

I wonder how much compilation time was improved after stopping including database.hh in compaction subsystem.

> 
> That is another header file - a rather silly one at that (how many source files need to handle IP addresses? Why does handling IP addresses take 2 seconds to compile?) included 291 times and single-handedly responsible for 3.7% of the build time. Heck, even seastar/core/future takes less of the build time (only slightly less, though).

**Comment by nyh** [comment]
I'm now focusing on one header file `<gms/inet_address.hh>`, and the story is very interesting.
It is included directly or indirectly by almost every source file (291 files) and this can't easily change because we use IP addresses as keys for a lot of things in Scylla. But the problem is not that we include it a lot but the fact it takes a whopping *2 seconds* in dev build mode for each of those includes. That's 600 seconds - 10 minutes of CPU time time - just for this silly header file defining an IP address. This can't make sense.

I'm now in the process of preparing Scylla and (unfortunately) Seastar patches to try to fix this situation. We should eventually do this for all "expensive" header files that ClangBuildAnalyzer reports.

**Comment by eliransin** [comment]
@nyh, when I have time I am planning to try the following, clang has this tool called iwyu (include what you use). It helps you to 
only include the header files you actually use. It removes (or at least tell you about) anything that can be "solved" by forward 
declarations. If my memory serves me right it even put a comment near every include what it is used for. This can help with 
deciding if some include files needs to be broken down.

**Comment by raphaelsc** [comment]
perhaps we should split database.hh, with table moved into table.hh. there are a few relevant files which include table, but not database, follow: ['alternator/auth.cc', 'alternator/executor.cc', 'api/storage_proxy.cc', 'auth/service.cc', 'connection_notifier.cc', 'db/batchlog_manager.cc', 'db/consistency_level.cc', 'db/cql_type_parser.cc', 'dht/range_streamer.cc', 'gms/failure_detector.cc', 'gms/gossiper.cc', 'index/secondary_index_manager.cc', 'redis/mutation_utils.cc', 'redis/options.cc', 'service/paxos/paxos_state.cc', 'sstables/sstable_directory.cc', 'sstables/sstable_set.cc', 'sstables/sstables.cc', 'sstables_loader.cc', 'streaming/stream_session.cc', 'streaming/stream_transfer_task.cc']

**Comment by nyh** [comment]
By the way, whoever wants to work on this issue, please note that there are two directions we can work on, both of them equally important:
1. Speeding up a **full** build. The invention of "dev mode" was a good step, and my patches from today (https://github.com/scylladb/scylla/pull/9875) also went in this direction.
2. Speeding up **partial** builds during development by ensuring that most of the time when you modify one file, only a handful of other files need to be recompiled - not the entire project. The worst thing for this goal is to have most of your code in header files included by many other files - so when you need modify some header code, half the world needs to be recompiled.

Some things, like templates and template functions for everything, cause problems for both goals. Some solutions - like type erasure (std::function et al.), pimpl, etc. - are helpful for both. Some of these solutions hurt (a bit) the resulting code's performance, so one of the nice things that Avi did in the past is to enable some of these compile-time-saving solutions just on "dev" build mode, and still do the slow but optimal thing in release mode. But in many non-critical parts of the code, the less optimal version is good enough for all build modes.

**Comment by avikivity** [comment]
> perhaps we should split database.hh, with table moved into table.hh. there are a few relevant files which include table, but not database, follow: ['alternator/auth.cc', 'alternator/executor.cc', 'api/storage_proxy.cc', 'auth/service.cc', 'connection_notifier.cc', 'db/batchlog_manager.cc', 'db/consistency_level.cc', 'db/cql_type_parser.cc', 'dht/range_streamer.cc', 'gms/failure_detector.cc', 'gms/gossiper.cc', 'index/secondary_index_manager.cc', 'redis/mutation_utils.cc', 'redis/options.cc', 'service/paxos/paxos_state.cc', 'sstables/sstable_directory.cc', 'sstables/sstable_set.cc', 'sstables/sstables.cc', 'sstables_loader.cc', 'streaming/stream_session.cc', 'streaming/stream_transfer_task.cc']

Many of these have nothing to do with the replica and should use data_dictionary instead.

**Comment by nyh** [comment]
I know this is a completely different project, in a different language, but it's still inspiring to read Ingo Molnar post from last month on his project spanning more than a year and 2297 (!) patches to make the Linux kernel compile 78% faster by completely reworking the header files. He notes that the first few - or even first 1500 (!) patches - give only small compile-speed improvements, after you're done with all the cleanup and reorganization, suddenly you can get huge compilation speed improvements from the next patches.

https://lwn.net/ml/linux-kernel/YdIfz+LMewetSaEB@gmail.com/

**Comment by nyh** [comment]
Issue #12962 proposes that build speed can be improved by using the new C++ "modules" feature instead of old-style header files.

**Comment by ptrsmrn** [comment]
Just an idea: how about using pre-compiled headers or C++20's modules in order to not compile some parts of the code over and over again?

**Comment by denesb** [comment]
> Just an idea: how about using pre-compiled headers or C++20's modules in order to not compile some parts of the code over and over again?

@tchaikov is working exactly on that (C++20 modules).

**Comment by michoecho** [comment]
> Just an idea: how about using pre-compiled headers or C++20's modules in order to not compile some parts of the code over and over again?

@ptrsmrn The comment right above yours says:

> Issue https://github.com/scylladb/scylladb/issues/12962 proposes that build speed can be improved by using the new C++ "modules" feature instead of old-style header files.

We were eager to use C++20 modules since their inception. We were doing first experiments 2 years ago. But the bottleneck is toolchain support for modules.

For example, it took until clang 15 (released last september) until we could compile a modules program that isn't miscompiled and segfaults on entry. Only clang 17 (not released yet) will have support for module dependency discovery ([p1689r5](https://wg21.link/p1689r5)). Cmake support was added in the last months. And so on.

AFAIK the state of modules in Scylla is that @tchaikov converted Seastar to a module (this is hidden behind an `ifdef`), and that work has been already merged to Seastar for quite some time. I'm not sure what are the next steps. Maybe we are waiting for clang 17 to be released?

**Comment by nyh** [comment]
> We were eager to use C++20 modules since their inception. We were doing first experiments 2 years ago. But the bottleneck is toolchain support for modules.

Issue https://github.com/scylladb/scylladb/issues/12962 was created specifically to track this effort.
@tchaikov would be nice to update that issue with more details on the current state and what it is that is holding us back today.

**Comment by ptrsmrn** [comment]
> I don't know what we can do about those. Maybe it makes sense to have external instantiation (only instantiating a specific version of a template) for some of these. Note that 138 seconds is almost 1% of the entire build time (around 15,000 seconds) - just for compiling this specific template 351 times :-(

@nyh I'm trying to push `extern template` in https://github.com/scylladb/scylladb/pull/16723#discussion_r1591104971

**Comment by lersek** [comment]
On a tangent: `ninja` does not deal well with `SIGSTP` (that is, when you press `^Z` on the terminal). References: https://github.com/ninja-build/ninja/issues/1241, https://github.com/ninja-build/ninja/pull/744.

(Mentioning this here because the longer the build time, the more distracting that ninja bug is -- if I have something in my shell command history that I want to reuse in a different terminal, or else if I need the CPU for some other "intensive" thing, temporarily, then I suspend the ninja build, and then it's quite surprising that the 20-ish clang processes continue hogging the laptop.)

**Comment by nyh** [comment]
This is just a feeling and not an accurate measurement - but I feel like the situation of the build time has gotten even worse recently.

I just did big but still partial build (after a "git pull", so no seastar rebuild and not everything had to be recompiled - but most did; I didn't compile tests at all, just the Scylla executable; just "dev" build mode) on my 6-core 12-hyperthread laptop in "performance" mode, and it took a whopping 26 minutes.

@scylladb/releng it would have been nice if we could make a long-term graph of **dev mode** full compilation time (not including **running** tests) to catch when we suffer compilation-time regressions. I doubt this will help us improve the situation, but at least it will warn us if we do something that makes the situation even worse.

Note that I believe that the compilation time of **dev build mode** is the most important metric, as this is what developers are expected to experience over and over. The build time of other modes also matter, but only as costs - we pay to Amazon for this build time, and I believe it's roughly 2 cents per hyperthread per hour, so 26 minutes of 12 hyperthreads will cost roughly 12 cents just for the dev build - and more for the entire build of all three modes.

**Comment by mykaul** [comment]
@nyh - did you use the Scylla optimized clang?

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh) - did you use the Scylla optimized clang?

No, I'm developing with what I have on my development machine, not dbuild or pieces of dbuild. How much faster have you seen Scylla's "optimized clang" to be over Fedora's clang?

If it's so great, why don't we offer the tricks we used to Fedora? I can hardly believe that it's only optimized for Scylla and not for other large buiilds.

**Comment by mykaul** [comment]
> > [@nyh](https://github.com/nyh) - did you use the Scylla optimized clang?
> 
> No, I'm developing with what I have on my development machine, not dbuild or pieces of dbuild. How much faster have you seen Scylla's "optimized clang" to be over Fedora's clang?
> 
> If it's so great, why don't we offer the tricks we used to Fedora? I can hardly believe that it's only optimized for Scylla and not for other large buiilds.

It was actually trained (PGO-wise) on compiling Scylla, so it is tailored to build Scylla faster.
I've heard 30% improvement, I would not dismiss it.

**Comment by radoslawcybulski** [comment]
@nyh i second your feeling - i went for PTO and suddenly scylla seems to be building much slower. I can try to hack precompiled header with std / boost / seastar and see if it makes tangible improvement. Unless it was tried without success?

**Comment by michoecho** [comment]
> [@nyh](https://github.com/nyh) i second your feeling - i went for PTO and suddenly scylla seems to be building much slower. I can try to hack precompiled header with std / boost / seastar and see if it makes tangible improvement. Unless it was tried without success?

It wasn't tried.

**Comment by avikivity** [comment]
I don't think precompiled headers work with distcc/ccache.

In C++26 precompiled headers for std were formalized: `import std;` but again won't work with distcc/ccache.

I believe most of the preprocessing work comes from our own code, not std/boost.

**Comment by radoslawcybulski** [comment]
https://ccache.dev/manual/3.2.4.html#_precompiled_headers
ccache supports precompiled headers, but needs some minor setup. I've a patch and i'm experimenting with it.

distcc doesn't seem to support it directly, but quick search tells me it can be worked around (supposedly distcc can be forced to compile header on each node and use it after).

Btw, full rebuild takes 33 minutes on my laptop. Granted i use -j10 for ninja.

**Comment by avikivity** [comment]
People should use build machines rather than laptops, I can't imagine that you did something so bad to deserve this punishment.

**Comment by nyh** [comment]
> People should use build machines rather than laptops, I can't imagine that you did something so bad to deserve this punishment.

This is easier said than done:
1. Do you have a **working** build machine setup in the office (we used to have one, but at some point it disappeared) for people in the office? Is it documented somewhere?
2. People who work remotely cannot effectively use distcc to the office build machines (too much bandwidth required), though can use a very-complex-to-set-up remote build (where the files remain on the remote computer). Maybe somebody should be assigned to prepare and document such a setup for all employees? IT shouldn't be just about security-theater fights :-)
3.  Finally, if talking about build machines at individual's homes (not offices), well, a (physically) huge build machine is inappropriate for most home setups - maybe somebody should be assigned to prepare an order for physically-small-and-not-noisy-but-strong home build machines. Pre-covid I actually had such small machines and used them, but they died (I guess from Covid ;-)) and were replaced by a huge machine that was inappropriate for my home setup so I just had to give it up without using it.

**Comment by michoecho** [comment]
> I believe most of the preprocessing work comes from our own code, not std/boost.

Too much believing, not enough knowing.

I _believe_ that churning through system headers is a big fraction of work for small `.cc` files, but a small fraction of work for the big `.cc` files.

> In C++26 precompiled headers for std were formalized: import std

`import std` was added in C++23. But support is still rough. (E.g. support in cmake is still experimental, cmake can't find the system module files on some distributions (like my nixos), it has the problems modules have in general, and so on).

**Comment by nyh** [comment]
> Btw, full rebuild takes 33 minutes on my laptop. Granted i use -j10 for ninja.

Reminds me about "Gates' Law" :-) 10 years after I opened this issue, although computers became faster building Scylla now takes 4 times longer (on these faster computer). It's not surprising that the amount of code grows as the project progresses, but we need to find a way to counteract this growth by more efficient compilation.

**Comment by nyh** [comment]
> > I believe most of the preprocessing work comes from our own code, not std/boost.
> 
> Too much believing, not enough knowing.
> 
> I _believe_ that churning through system headers is a big fraction of work for small `.cc` files, but a small fraction of work for the big `.cc` files.

This issue now has hundreds of comments, some of them tested all sorts of beliefs.

One of the most interesting observations that we made in the past and I believe is still very accurate, is that compilation time is proportional to the generated **object file**'s size (.o) - when we find a technique that halves the generated object size, it also halves the compilation time. This is surprising - after all, "churning through system headers" does a lot of work but generates no output to the object file, right? Well, it seems the conclusion is that "churning through system headers" doesn't take a high percentage of our build time - what is most heavy is generating templated code and optimizing this code (in modern C++, we can't really "disable optimization" completely any more). You can see many examples above where a 1000 line file takes a minute to build - not because 1000 or even 100,000 lines take a minute to read but because of the huge amount of code generated from templates used in those 1000 lines.

**Comment by nyh** [comment]
> `import std` was added in C++23. But support is still rough. (E.g. support in cmake is still experimental, cmake can't find the system module files on some distributions (like my nixos), it has the problems modules have in general, and so on).

And we must use cmake because... what?

**Comment by avikivity** [comment]
> > `import std` was added in C++23. But support is still rough. (E.g. support in cmake is still experimental, cmake can't find the system module files on some distributions (like my nixos), it has the problems modules have in general, and so on).
> 
> And we must use cmake because... what?

The reason to use cmake is to reduce the amount of gnarly python code by standardized cmake modules.

In practice cmake was so inflexible most of the gnarly python code was replaced by gnarly cmake code.

Another reason was that cmake has the ability to integrate nested cmake modules (seastar, abseil) into a single build plan, but I think there were some problems with that.

**Comment by michoecho** [comment]
> > `import std` was added in C++23. But support is still rough. (E.g. support in cmake is still experimental, cmake can't find the system module files on some distributions (like my nixos), it has the problems modules have in general, and so on).
> 
> And we must use cmake because... what?

I wasn't making an argument about using std modules in Scylla specifically, I was just loosely commenting on the overall state of module support by various tools. I'm not a proponent of cmake.

@avikivity Now that the only cmake person (Kefu) left the project, what's the state of "cmake vs configure.py" anyway?

**Comment by radoslawcybulski** [comment]
I'm using CMake and it works. Granted i call `configure.py --use-cmake`.

**Comment by radoslawcybulski** [comment]
Here is a patch for precompiled headers.

https://github.com/scylladb/scylladb/pull/25182

I left them on by default, but given the distcc situation it might be not feasible, what do you think?

**Comment by avikivity** [comment]
> Here is a patch for precompiled headers.
> 
> [#25182](https://github.com/scylladb/scylladb/pull/25182)
> 
> I left them on by default, but given the distcc situation it might be not feasible, what do you think?

Please maintain parity between cmake and configure.py until we make the switch.

We can enable it conditionally for distcc non-users.

**Comment by avikivity** [comment]
> > > `import std` was added in C++23. But support is still rough. (E.g. support in cmake is still experimental, cmake can't find the system module files on some distributions (like my nixos), it has the problems modules have in general, and so on).
> > 
> > 
> > And we must use cmake because... what?
> 
> I wasn't making an argument about using std modules in Scylla specifically, I was just loosely commenting on the overall state of module support by various tools. I'm not a proponent of cmake.
> 
> [@avikivity](https://github.com/avikivity) Now that the only cmake person (Kefu) left the project, what's the state of "cmake vs configure.py" anyway?

It needs someone to prove we don't lose (or gain) any compilation flags, then we can retire configure.py.

**Comment by michoecho** [comment]
> > Here is a patch for precompiled headers.
> > [#25182](https://github.com/scylladb/scylladb/pull/25182)
> > I left them on by default, but given the distcc situation it might be not feasible, what do you think?
> 
> Please maintain parity between cmake and configure.py until we make the switch.

@avikivity For what? What's wrong with having a cmake-only QoL improvement?

> We can enable it conditionally for distcc non-users.

How do you know if distcc is used or not?

**Comment by radoslawcybulski** [comment]
@avikivity ok, i will do it. Are we talking about compilation flags for object file / linkage or something else on top of that?

**Comment by avikivity** [comment]
> > > Here is a patch for precompiled headers.
> > > [#25182](https://github.com/scylladb/scylladb/pull/25182)
> > > I left them on by default, but given the distcc situation it might be not feasible, what do you think?
> > 
> > 
> > Please maintain parity between cmake and configure.py until we make the switch.
> 
> [@avikivity](https://github.com/avikivity) For what? What's wrong with having a cmake-only QoL improvement?
> 


I'm not taking the python->cmake transition for granted, I'll believe it when I see it.

> > We can enable it conditionally for distcc non-users.
> 
> How do you know if distcc is used or not?

Since distcc itself knows if its used or not, there has to be a solution. Like looking if distcc is installed, has configuration files/environment, and is not disabled.

**Comment by avikivity** [comment]
> [@avikivity](https://github.com/avikivity) ok, i will do it. Are we talking about compilation flags for object file / linkage or something else on top of that?

Compilation flags for representative object files (service/storage_proxy.o / core/reactor.o) in debug/dev/release modes, link flags for iotune/scylla in release mode. You can sort+diff them, though order matters when flags negate each other.

To be extra sure, run a refresh-pgo-profiles.sh cycle, verify new profiles generated in pgo/profiles/*, and verify perf-simple-query --smp 1 has not regressed when comparing instructions per op.

**Comment by mykaul** [comment]
## Build Speed Improvement Plan — Static Analysis Findings & Prioritized Actions

**Goal:** Achieve ~30% faster clean dev builds (scylla binary + all tests, Seastar/Abseil cached).

Inspired by [ClickHouse's compilation speed initiative](https://github.com/ClickHouse/ClickHouse/issues/96721), we performed a thorough static analysis of ScyllaDB's build structure. Below are the findings and a prioritized action plan.

---

### Codebase Build Profile

| Metric | Value |
|--------|-------|
| .cc files (ScyllaDB-owned) | ~852 |
| .hh files (ScyllaDB-owned) | ~1,166 |
| TUs for scylla binary | ~594 |
| TUs for tests | ~302 |
| Total TUs (dev build) | ~896 |
| Link targets | ~235 (38 static archives + 197 executables) |
| PCH (`stdafx.hh`) | 376 includes (shared by ~33 library targets) |
| Parallel link jobs | 15-16 (4GB/job on 64GB RAM) |

**Build system:** CMake + Ninja, C++23, Clang, lld  
**Dev mode:** `-O2`, no debug info, no LTO, no sanitizers, Seastar as shared library  
**No build profiling exists today** (no `-ftime-trace`, no ClangBuildAnalyzer)

---

### TIER 1 — High-Value Changes

#### 1. Add `-ftime-trace` Build Profiling Infrastructure

Before making changes, we need measurement. Add a CMake option to enable Clang's `-ftime-trace` flag and integrate [ClangBuildAnalyzer](https://github.com/ArasLabs/ClangBuildAnalyzer) for analysis.

**Impact:** Enables data-driven decisions for all subsequent optimizations.

---

#### 2. Remove `storage_service.hh` from `storage_proxy.hh` _(trivial fix)_

**File:** `service/storage_proxy.hh` line 42

`storage_proxy.hh` includes `storage_service.hh` but does NOT reference `storage_service` at all. The types it actually needs (`fencing_token`, `topology_state_machine`) already arrive via `abstract_replication_strategy.hh → token_metadata.hh → topology_state_machine.hh`. The only "missing" type is `raft_group0_client&` (one reference parameter at line 567) — trivially forward-declarable.

**Fix:** Delete one `#include` line, add one forward declaration.

**Impact:** 74 TUs shed the heavy `storage_service.hh` chain (`db/system_keyspace.hh`, `gms/gossiper.hh`, QoS/auth, `mutation/mutation.hh`, `locator/snitch_base.hh`, etc.).

---

#### 3. Pimpl `pluggable<storage_service>` in `migration_manager.hh` _(largest header win)_

**File:** `service/migration_manager.hh` line 21, 61

`migration_manager.hh` needs `storage_service.hh` because `utils::pluggable<storage_service> _ss` (member at line 61) has inline methods requiring the complete type.

**Fix:** Replace with `std::unique_ptr<utils::pluggable<storage_service>> _ss`, define destructor in `.cc`.

**Impact:** This is the **single largest header chain elimination** — `migration_manager.hh` has NO other path to `token_metadata.hh`, `tablets.hh`, `topology_state_machine.hh`. **58 .cc files** (mostly CQL DDL statements in `cql3/statements/`) lose the entire topology/tablets/gossiper/system_keyspace/QoS/auth/mutation chain (~5,000+ lines each).

---

#### 4. Split `storage_proxy.cc` into 3 Files

**File:** `service/storage_proxy.cc` — 7,493 lines, only 1 tiny anonymous namespace (24 lines)

Clear functional boundaries:
- `remote` inner class (~1,047 lines)  
- Read path (~2,500 lines)  
- Write + CAS path (remainder)

**Impact:** Better build parallelism — currently a serialization bottleneck on the critical path.

---

#### 5. Create `test-lib-minimal` for BOOST Tests

**File:** `test/CMakeLists.txt` lines 65-69

Currently ALL 195 test executables unconditionally link `test-lib`, which PUBLIC-links `scylla-main` + `replica`. But `test-lib`'s 2 PUBLIC source files (`boost_test_tree_lister.cc`, `boost_tree_lister_injector.cc`) have **zero** scylla-main dependencies — they're pure Boost.Test infrastructure.

**37 of 52 BOOST tests** specify no extra LIBRARIES and don't need scylla-main.

**Fix:** Create `test-lib-minimal` with only Boost.Test dependencies, use it for simple BOOST tests.

**Impact:** 37 tests get trivial link steps instead of linking the entire ScyllaDB binary.

---

#### 6. Split `repair/row_level.cc` and Other Large TUs

**File:** `repair/row_level.cc` — 4,047 lines, 3-file library

Contains `repair_meta` class (1,481 lines, defined in `.cc` at line 752). Extract to internal header, split into 2-3 files.

Other candidates for assessment: `alternator/executor.cc` (6,214 lines), `replica/table.cc` (5,647 lines), `service/storage_service.cc` (6,579 lines).

**Impact:** Better build parallelism for the repair library (currently only 3 source files).

---

### TIER 2 — Medium-Value Changes

#### 7. Create `seastar_fwd.hh`

Create a ScyllaDB-owned forward-declaration header: `namespace seastar { template <typename Service> class sharded; }` (precedent: `seastarx.hh` already forward-declares `seastar::shared_ptr`).

Of ~91 `.hh` files including `sharded.hh`, **48-50 MUST keep** the full include (inherit `peering_sharded_service`/`async_sharded_service`, by-value members, inline `.local()` calls). Only **~41-43 files** are candidates.

#### 8. Create Forward-Declaration Headers

- `storage_service_fwd.hh`
- `system_keyspace_fwd.hh`  
- `gossiper_fwd.hh`
- `config_fwd.hh`

These heavy service headers are included by many files that only need a forward declaration.

#### 9. Reduce `storage_service.hh` Weight

After items #2 and #3 eliminate the two biggest unnecessary inclusion paths, audit `storage_service.hh`'s 40 includes for further forward-declaration opportunities.

#### 10. Move More Tests into `combined_tests`

Currently 71 of 143 boost tests are in `combined_tests`. Moving more standalone tests reduces the number of link targets.

#### 11. Remove `boost/test/unit_test.hpp` from `stdafx.hh`

Test-only header included in the production PCH at line 70.

---

### What Was Ruled Out

- **`locator/tablets_fwd.hh`:** Tablet types are small value types used by value in almost all includer `.hh` files. Only ~2-4 of 20 could use forward declarations — not worth the churn.
- **`types/types.cc` split:** 13 anonymous namespaces (930 lines, 23% of file), 75 static functions, cross-cutting visitor pattern — impractical to split.
- **`<functional>` removal:** Seastar's `future.hh`/`sstring.hh` force it everywhere.
- **`exceptions/exceptions.hh` cleanup:** Already lightweight.
- **Seastar subtree modifications:** Out of scope.
- **ANTLR parser generation:** Out of scope.

---

### Estimated Impact

| Scenario | Estimated Improvement |
|----------|----------------------|
| Conservative (Tier 1 only, partial) | ~14-18% |
| Moderate (Full Tier 1 + some Tier 2) | ~22-30% |
| Aggressive (All tiers, best case) | ~26-34% |

These estimates are from static analysis only. The `-ftime-trace` baseline (Item 1) will provide real measurements to validate and prioritize further.

---

### Execution Plan

- **Week 1:** Items 1-3 (profiling infrastructure + two header fixes + PCH cleanup)
- **Week 2:** Items 4-6 (TU splits + test infrastructure)
- **Week 3+:** Tier 2 items based on profiling data

All changes maintain full backward compatibility. Each item is independently valuable and can be merged separately.

**Comment by mykaul** [comment]
@nyh - see above. I think I'll try with AI to do some build time reduction. It's a long plan.

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh) - see above. I think I'll try with AI to do some build time reduction. It's a long plan.

I'm happy that the AI is picking this up. I agree with its conclusion that the first step should be to use ClangBuildAnalyzer - I also recommended this four years ago, see https://github.com/scylladb/scylladb/issues/1#issuecomment-1004106087, and used it for a few small changes. 

I don't understand why the AI continued to come up with suggestions before getting the results from ClangBuildAnalyzer, which it itself said should be the first thing to do. I think you should ask it to come up with the plan after it gets the measurements, not before, no? Although maybe some of its ideas won't change after getting more input.

I recently asked AI (with my supervision) do some improvements in this area and the result was https://github.com/scylladb/scylladb/pull/28992 and https://github.com/scylladb/scylladb/pull/29009 - but it is almost hopeless to get those changes reviewed or merged (we've been in almost constant mix of merge freeze and important-urgent-things-for-release for more than a month now). https://github.com/scylladb/scylladb/pull/28992 changes config.hh which was one of the things your AI suggested too.

Anyway, if AI is going to make these suggestions I suggest that it should consider the following guidelines:
1. Every change should explain whether it is trying to improve full-build time (total build of everything) or incremental-build time (changing one line of code causes 100 files to need to be built - or not). Both are important, but are different.
2. Every change should either make a proven benefit (e.g., 1% faster build time, or 1% smaller object files) or if it's not, and it's just a buildup towards another change that will bring some improvements - explain how you know that.
3. Massive reorganizations get negative points. So do ugly arbitrary changes (e.g., split a long file randomly into three parts, without thinking how to make the code nicely organized - not just smaller pieces).
4. Some changes like PIMPL or type erasure make performance worse but can bring very large benefits. If the type involved is important for performance, consider doing the change only for dev build mode (we have such things in our code already). But if the type is not important for performance it can be done for all build modes, increasing the build speed of all build modes not just dev build mode.

**Comment by mykaul** [comment]
## Idea: Disable `SEASTAR_LOGGER_COMPILE_TIME_FMT` in dev builds only

**Data from `-ftime-trace` profiling of a clean dev build (601 TUs, 24,360s total compile time):**

`fmt::` template instantiation accounts for ~4,900s (20% of total build time). Of that, **~2,315s (47% of fmt cost, 9.5% of total build)** is spent on `fstring<>` + `format_string_checker<>` — the compile-time format string validation enabled by `SEASTAR_LOGGER_COMPILE_TIME_FMT`.

This is `consteval` overhead: it produces zero code, it's pure frontend compile-time cost. Every `logger.info("...", args...)` and `fmt::format("...", args...)` call triggers it. There are ~5,000 such calls across the codebase.

Disabling `SEASTAR_LOGGER_COMPILE_TIME_FMT` for dev mode would switch logging to `fmt::runtime()`, deferring format string validation to runtime. Estimated savings: **~2,315s cumulative → ~145s wall-clock at -j16 (~9.5% of build)**.

**Tradeoff:** Format string mismatches (wrong number/type of args) become runtime errors instead of compile errors in dev builds. Release/debug builds would retain compile-time checking.

This is a policy decision, not a code change — posting here for discussion.

**Comment by avikivity** [comment]
Disabling compile-time format checking is reasonable, if we can prove it actually saves so much.

**Comment by nyh** [comment]
> Disabling compile-time format checking is reasonable, if we can prove it actually saves so much.

Exactly my thinking:

1. It's perfectly fine to disable compile-time format checking from dev mode, or even all modes except release. These compile-time format checks are rarely helpful for developers (you don't need to check old code, and the new code you write - you run and will see the runtime problems), but in any case CI will still show you these bugs - just like it runs debug builds for you.
2. I don't trust the AI's estimation that this will save 9.5% of the build time. This is a **huge**, surprising, saving. I think we should measure it to confirm it's not an halucination. I personally think that even if it's 1% saving it's worth it for such a trivial patch, but I think we should know.

**Comment by mykaul** [comment]
It will eventually boil down to splitting large source files to 2-3 smaller ones. Whether this is the right solution, generally, or per each file, remains to be seen.
It seems to help substantially, but we'll see.

**Comment by nyh** [comment]
> It will eventually boil down to splitting large source files to 2-3 smaller ones.

You're suggesting that we can solve every one - or even just most - of the compilation slowness problems by splitting large files? This has not been my experience:

Splitting a large source file to 3 has an big effect when it's a "straggler" (the last file to be compiled) or when it's the only file that needs to be recompiled, but almost zero benefit in most other cases. It can also have downsides - it can increase total compilation time if all 3 chunks need to include the same files and compile the same templates, and unless done with some understanding of what to put where can make the code harder to understand or find.

In https://github.com/scylladb/scylladb/pull/29046 I split the 7000-line alternator/executor.cc into three files, so I do think this is an important thing to do. But I don't think it will solve all our problems. Not even most of them.

> Whether this is the right solution, generally, or per each file, remains to be seen. It seems to help substantially, but we'll see.

Can you please give a concrete example where it helps substantially?

**Comment by mykaul** [comment]
> > It will eventually boil down to splitting large source files to 2-3 smaller ones.
> 
> You're suggesting that we can solve every one - or even just most - of the compilation slowness problems by splitting large files? This has not been my experience:
> 
> Splitting a large source file to 3 has an big effect when it's a "straggler" (the last file to be compiled) or when it's the only file that needs to be recompiled, but almost zero benefit in most other cases. It can also have downsides - it can increase total compilation time if all 3 chunks need to include the same files and compile the same templates, and unless done with some understanding of what to put where can make the code harder to understand or find.
> 
> In [#29046](https://github.com/scylladb/scylladb/pull/29046) I split the 7000-line alternator/executor.cc into three files, so I do think this is an important thing to do. But I don't think it will solve all our problems. Not even most of them.
> 
> > Whether this is the right solution, generally, or per each file, remains to be seen. It seems to help substantially, but we'll see.
> 
> Can you please give a concrete example where it helps substantially?

Random example (out of several):
```
    replica: split database.cc into core + operations TU for build parallelism
    
    Split replica/database.cc (3,991 lines, ~190s compile) into two TUs:
    - database.cc (2,623 lines): init, schema, query, apply
    - database_operations.cc (1,430 lines): lifecycle (start/shutdown/stop),
      flush, truncate, snapshot, drain, tables_metadata methods, multishard
      streaming reader, query_mutations/query_data free functions, QoS
      callbacks, RF rack validation, sample_data_files
    
    Cross-TU dependencies resolved via extern logger declaration (dblog)
    and duplication of trivial is_system_table helper.
    
    Compile time improvement (measured with -ftime-trace, -j2):
    - Before: ~190s (single TU)
    - After: database.cc 23.5s + database_operations.cc 21.6s = 45.1s total
    - CPU reduction: ~145s (76%)
```

I'm not convinced of the outcome - yet. I will ask for full, end-to-end compilation and linking results and so forth, but this is an example.
I'm not sure it's bad, btw, regardless. I think we might have grown some modules to be too big in general.

**Comment by mykaul** [comment]
Interim status:
```
2026-04-14 13:00 +0300 Yaniv Michael Kaul                      o [ykaul/remove-storage-service-from-proxy] tools: split scylla-sstable.cc dump/inspect operations into separate file for build parallelism
2026-04-14 11:49 +0300 Yaniv Michael Kaul                      o service: split topology_coordinator.cc into migration engine + transitions for build parallelism
2026-04-14 11:23 +0300 Yaniv Michael Kaul                      o repair: split row_level.cc into protocol core + orchestration for build parallelism
2026-04-14 10:08 +0300 Yaniv Michael Kaul                      o sstables: split sstables.cc query/streaming code into separate file for build parallelism
2026-04-14 08:55 +0300 Yaniv Michael Kaul                      o api: split storage_service.cc routes into separate TU for build parallelism
2026-04-14 08:55 +0300 Yaniv Michael Kaul                      o replica: split database.cc into core + operations TU for build parallelism
2026-04-14 05:55 +0300 Yaniv Michael Kaul                      o service: split storage_proxy.cc read path into separate TU for build parallelism
2026-04-14 04:49 +0300 Yaniv Michael Kaul                      o service: split storage_service.cc into 4 TUs for build parallelism
2026-04-14 03:22 +0300 Yaniv Michael Kaul                      o message: split messaging_service.cc into base + verbs TU for build parallelism
2026-04-13 23:27 +0300 Yaniv Michael Kaul                      o mutation: remove idl/mutation.dist.impl.hh from mutation_partition_view.hh
2026-04-13 23:26 +0300 Yaniv Michael Kaul                      o mutation: move frozen_mutation::printer formatter body to .cc file
2026-04-13 19:20 +0300 Yaniv Michael Kaul                      o db, test: add missing includes exposed by storage_proxy.hh cleanup
2026-04-13 18:51 +0300 Yaniv Michael Kaul                      o service: remove unnecessary storage_service.hh include from storage_proxy.hh
```

(AI is still crunching and killing my poor laptop. Probably will take few more days. NO conclusion yet)

**Comment by nyh** [comment]
> > Can you please give a concrete example where it helps substantially?
> 
> Random example (out of several):
> 
> ```
>     replica: split database.cc into core + operations TU for build parallelism
>     
>     Split replica/database.cc (3,991 lines, ~190s compile) into two TUs:
>     - database.cc (2,623 lines): init, schema, query, apply
>     - database_operations.cc (1,430 lines): lifecycle (start/shutdown/stop),
>       flush, truncate, snapshot, drain, tables_metadata methods, multishard
>       streaming reader, query_mutations/query_data free functions, QoS
>       callbacks, RF rack validation, sample_data_files
>     
>     Cross-TU dependencies resolved via extern logger declaration (dblog)
>     and duplication of trivial is_system_table helper.
>     
>     Compile time improvement (measured with -ftime-trace, -j2):
>     - Before: ~190s (single TU)
>     - After: database.cc 23.5s + database_operations.cc 21.6s = 45.1s total
>     - CPU reduction: ~145s (76%)
> ```

Are these measurements real?

I don't understand how they could be real. How can you split a file that takes 190s to compile into two files that take, serially, 45 seconds to compile? Is there some algorithm in the compiler which is quadratic in the number of lines in the file or something? I have never actually need this happening, it's very surprising.

The first line in the above blurb - "split database.cc into core + operations TU for build parallelism" is what **does** make sense - you can split a file that takes 2 minutes to compile into two files that each take 1 minute to compile, and compile them in parallel. However, this is not what the numbers you showed show - they claim that the **total** CPU time went down from 190s to 45s because we split the file into two. I don't have any idea how this is possible.

**Comment by mykaul** [comment]
> > > Can you please give a concrete example where it helps substantially?
> > 
> > 
> > Random example (out of several):
> > ```
> >     replica: split database.cc into core + operations TU for build parallelism
> >     
> >     Split replica/database.cc (3,991 lines, ~190s compile) into two TUs:
> >     - database.cc (2,623 lines): init, schema, query, apply
> >     - database_operations.cc (1,430 lines): lifecycle (start/shutdown/stop),
> >       flush, truncate, snapshot, drain, tables_metadata methods, multishard
> >       streaming reader, query_mutations/query_data free functions, QoS
> >       callbacks, RF rack validation, sample_data_files
> >     
> >     Cross-TU dependencies resolved via extern logger declaration (dblog)
> >     and duplication of trivial is_system_table helper.
> >     
> >     Compile time improvement (measured with -ftime-trace, -j2):
> >     - Before: ~190s (single TU)
> >     - After: database.cc 23.5s + database_operations.cc 21.6s = 45.1s total
> >     - CPU reduction: ~145s (76%)
> > ```
> 
> Are these measurements real?
> 
> I don't understand how they could be real. How can you split a file that takes 190s to compile into two files that take, serially, 45 seconds to compile? Is there some algorithm in the compiler which is quadratic in the number of lines in the file or something? I have never actually need this happening, it's very surprising.
> 
> The first line in the above blurb - "split database.cc into core + operations TU for build parallelism" is what **does** make sense - you can split a file that takes 2 minutes to compile into two files that each take 1 minute to compile, and compile them in parallel. However, this is not what the numbers you showed show - they claim that the **total** CPU time went down from 190s to 45s because we split the file into two. I don't have any idea how this is possible.

I don't trust the numbers either. I've asked it to complete the work and re-measure. It's based on -ftime-trace values measurements. I'll need end-to-end results.
(My hope is reduced complexity in compilation of smaller items, but I find it very hard to believe)

**Comment by mykaul** [comment]
## File-splitting benchmark results (April 2026)

Inspired by [ClickHouse#96721](https://github.com/ClickHouse/ClickHouse/issues/96721), I tried systematically splitting the slowest-compiling .cc files to improve build parallelism. 23 commits on branch `ykaul/remove-storage-service-from-proxy` (not pushed), splitting 18 files into ~42 files total.

### Per-file results looked amazing

| File | Before | After (critical path) | Reduction |
|---|---|---|---|
| messaging_service.cc | 508s | 53.6s | 89% |
| storage_service.cc | 371s | 45.8s | 88% |
| storage_proxy.cc | 259s | 32.6s | 87% |
| replica/table.cc | 125s | 28.7s | 77% |
| db/system_keyspace.cc | 147s | 28.9s | 80% |
| + 13 more splits | | | similar |

Total CPU compile time across 630 files dropped from ~24,000s to ~14,700s (39% reduction).

### But wall-clock tells a different story

Both runs: `-j16`, 16 cores, 30GB RAM, `--compiler-cache=none`, no `-ftime-trace`, Seastar/Abseil pre-cached, PCH rebuilt, identical machine:

| Metric | Baseline (master) | After (23 commits) | Delta |
|---|---|---|---|
| Wall-clock time | **24m07s** | **25m09s** | **+62s (4.3% slower)** |
| Ninja steps | 633 | 657 | +24 |

### Why it didn't help

1. **More files = more fixed overhead**: Each new .cc file pays a ~15s PCH/frontend tax (the 202MB PCH with 376 headers). 24 extra files × ~15s ≈ 360s extra CPU, or ~22s wall time at -j16.
2. **Parallelism was already sufficient**: With 600+ files and -j16, the machine was never truly idle waiting for a single slow file. The "long pole" files that took 500s were finishing while other work ran on other cores.
3. **Superlinear optimizer savings are real but offset by the frontend duplication**: Splitting a 143s file into two 16s files saves 111s CPU, but adds 15s of duplicated frontend work. At -j16 where cores are saturated, the net effect is negative.

### What this means

File splitting alone does not improve wall-clock build time on a 16-core machine for this codebase. The build is CPU-bound across all cores, not bottlenecked on individual slow files.

Potential paths that *might* help wall-clock:
- **Reduce PCH size** (the ~15s per-file frontend tax is the real enemy)
- **Extern template instantiations** for heavy templates like `fmt::format`, `loading_cache` (reduce CPU work per file without adding files)
- **De-templating** hot code paths (reduce template instantiation CPU across all files)
- **Higher core count machines** (the splits would help more on 64+ cores where the long-pole files do become bottlenecks)

The CPU time reduction is real (39%) and would show up as faster builds on machines with many more cores, or with compiler caching (sccache/ccache) where per-file time directly matters.

**Comment by nyh** [comment]
@mykaul I don't understand anything (and don't understand if you wrote any of this or it was direct AI output with no human touch).

First, it says that:

        messaging_service.cc 	508s 	53.6s

Which suggests that just for this single file, splitting it into 2 or 3 files, made it compile 10 times (!) faster. This is "amazing" as the text says - it doesn't even matter if it's 53.6s times two or three CPUs - it's still a huge win. Which doesn't make any sense, to be honest. I think it's just an hallucination - unless I misunderstand what these numbers mean.

Then, the final result:

         Wall-clock time 	24m07s 	25m09s

This is **exactly what I expected** and exactly my experience: Splitting files make the overall compilation slower, not faster.

In short, I can't make any sense of this (I am guessing) AI-written output. It sounds like the whole idea of splitting files doesn't work as a way to reduce the total build time! It can help other things like in case of stragglers (what the AI called "long-pole files do become bottlenecks") or modifying that very large file, but not total build time. And then the AI gives apologetics like "The CPU time reduction is real" and "would show up as faster builds on machines with many more cores". Ah?? 16 cores is not enough, and the nonsensical magical properties splitting individual files only show up on more cores? I can't make any sense of this. Again, I can understand how splitting a file in two can result in individual files that compile (in parllel) twice fast. I don't understand how it compiles 10 times faster.

**Comment by mykaul** [comment]
@nyh - I'm not sure what is misunderstood from the summary. File-splitting overall is ineffective. Apparently, the process of 'de-templating' which seemed to work well for Clickhouse was ineffective for us. Why? Unsure yet.

**Comment by nyh** [comment]
> [@nyh](https://github.com/nyh) - I'm not sure what is misunderstood from the summary. File-splitting overall is ineffective. 

I understood - and expected - that file-splitting overall is ineffective (at least - when **full** compilation time is concerned and you don't have a thousand cores). 
My problem was not with that conclusion, but with all that other text that the AI apparently wrote and makes no sense, like how splitting one file into two made compilation of that file 10 times (!) faster, or how compilation is now 39% faster but for `<reasons>` you can't measure it and it appears to be slower ;-)

**Comment by mykaul** [comment]
Here's a smaller, but tangible success:
```
    pch: remove seastar/http/api_docs.hh from precompiled header
    
    The api_docs.hh header contains inline method bodies (api_registry::handle)
    that call seastar::json::formatter::to_json(), forcing the compiler to
    instantiate seastar::json template specializations (json_list_template,
    formatter::write, do_with, etc.) in every compilation unit — even files
    that never use any HTTP/JSON API types.
    
    Measured ~6s of wasted template instantiation per file × ~620 files =
    ~3,700s total CPU. Only 2 files outside the PCH include api_docs.hh
    directly, so removing it has no impact on code that actually uses these
    types.
    
    Wall-clock build time (-j16, Seastar/Abseil cached):
      Before (with loading_cache fix): avg 23m29s
      After:                           avg 23m04s  (-1.8%)
      vs original baseline:            avg 24m01s  (-4.0%)
```

**Comment by nyh** [comment]
Very good. I don't know who or how decided what actually goes into this precompiled header. Maybe there are other files in there that should be there.

**Comment by mykaul** [comment]
@nyh - Q: wouldn't the split help nicely with incremental compilation? I was solely focused on clean -> full compilation, but maybe the interesting path to pursue is incremental compilation?
